nixos/modules/services/networking/bee.nix

```diff
@@ -0,0 +1,166 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+let
+  cfg = config.services.bee;
+in {
+  meta = {
+    # doc = ./bee.xml;
+    maintainers = with maintainers; [ attila-lendvai ];
+  };
+
+  ### interface
+
+  options = {
+    services.bee = {
+      enable = mkEnableOption "Start up Ethereum Swarm Bee";
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.bee;
+        defaultText = "pkgs.bee";
+        example = "pkgs.bee-unstable";
+        description = "The package providing the bee binary for the service.";
+      };
+
+      ### NOTE let's call the nix variables the same as bee calls them
+
+      config = mkOption {
```

Comment:
I would like to introduce you to [RFC42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md). That RFC might suggest you use something like this:

```
let
  format = pkgs.formats.yaml {};
  configFile = format.generate "bee.yaml" cfg.settings;
in
  options = {
    services.bee.settings = {
      type = format.type;
    };
  };
```

Feel free to ping me if you have questions, but the RFC does a really good job of explaining these ideas.

Reply:
i'll read through that RFC. meanwhile, it would help me a lot if you could point me to an example nix file that is done based on best practices, so that i can mimic what's there.

i was mimicing the bitcoin nix file when i was putting together this one.

i'm new to nix.

Reply:
> i'm new to nix.

Exciting! In that case I'll try to find some time later today to go through my comments in **way** more detail. Thanks for taking an interest in NixOS!

Reply:
@attila-lendvai I found some time and put this together: https://github.com/NixOS/nixpkgs/compare/master...aanderse:nixos/bee?expand=1

Let me know what you think.

Reply:
@aanderse thank you for your work! it looks good to my uninitiated eyes. i've pushed them into this PR.

can you please help me with how i can test this on my nixos install? the manual is either out of date, or i just couldn't figure out how to install this service onto my nixos install.

there's quite some work left to do, because ethereum's clef service needs to be packaged/started with a special setup that is suitable for bee (relaxed permissions due to their automated interactions).

i think it doesn't make sense to package a bee-clef separately from bee, but i asked for clarification whether bee-clef is ever useful without bee also being installed. if not, then i think the service should depend on the go-ethereum package to provide the clef binary, and set up a systemd service with the proper config for bee.

PS: i first installed nixos a couple of weeks ago as my main linux, and i'm loving it! nix and nixos has its quirks, especially in some of the opaque DWIM behavior that is rather hard to follow for a newcomer... and the naming of the entities in the code is very much not helpful. related parts of the codebase should be grep'able to help newcomers grasp it, and it's often not the case.

Reply:
If you're looking to test this against your current system and assuming you have a local checkout of nixpkgs under `/home/attila-lendvai/nixpkgs` you could temporarily add something (very impure) like this to your `configuration.nix` and run `nixos-rebuild test`:
```
{ config, pkgs, ... }:
let
  master = import /home/attila-lendvai/nixpkgs {};
in
{
  imports = [
    /home/attila-lendvai/nixpkgs/nixos/modules/services/networking/bee.nix
  ];

  nixpkgs.overlays = [
    (self: super: {
      bee = (master.callPackage /home/attila-lendvai/nixpkgs/pkgs/applications/networking/bee.nix { }).bee;
    })
  ];

  services.bee.enable = true;
}
```

That probably isn't the best piece of advice I can give you, though. `build-vm` is your friend.